http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java

```diff
@@ -1805,6 +1871,63 @@ HttpToHttp2ConnectionHandlerBuilder newHttp2ConnectionHandlerBuilder(
 
     }
 
+    private Flux<io.micronaut.http.HttpResponse<Object>> readBodyOnError(Argument<?> errorType, Flux<io.micronaut.http.HttpResponse<Object>> publisher) {
+        if (errorType != null && errorType != HttpClient.DEFAULT_ERROR_TYPE) {
+            return publisher.onErrorResume(clientException -> {
+                if (clientException instanceof HttpClientResponseException) {
+                    final HttpResponse<?> response = ((HttpClientResponseException) clientException).getResponse();
+                    if (response instanceof NettyStreamedHttpResponse) {
+                        return Mono.create(emitter -> {
+                            NettyStreamedHttpResponse streamedResponse = (NettyStreamedHttpResponse<?>) response;
+                            final StreamedHttpResponse nettyResponse = streamedResponse.getNettyResponse();
+                            nettyResponse.subscribe(new Subscriber<HttpContent>() {
+                                final CompositeByteBuf buffer = byteBufferFactory.getNativeAllocator().compositeBuffer();
+                                Subscription s;
+                                @Override
+                                public void onSubscribe(Subscription s) {
+                                    this.s = s;
+                                    s.request(1);
+                                }
+
+                                @Override
+                                public void onNext(HttpContent httpContent) {
+                                    buffer.addComponent(true, httpContent.content());
+                                    s.request(1);
+                                }
+
+                                @Override
+                                public void onError(Throwable t) {
+                                    emitter.error(t);
```

Comment:
Seems the buffer cleanup logic would need to occur `onError` as well since `onComplete` may not be called?

Reply:
Right on